### PR TITLE
Less docker magic in docs building

### PR DESCRIPTION
**Pull request by mik-laj**
Recently @potiuk worked on building docs in parallel and managed to do so, but used a lot of Docker tricks. https://github.com/apache/airflow/blob/master/docs/build_docs.py

I would like to present another workaround that patches only one function from `sphinx-autoapi` and does not change the rest of the process. This simplifies the build process and makes it independent of Docker.  You can use Docker to run a documentation build if you want, but it's not required.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: docs/exts/docs_build/run-patched-sphinx.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                "AutoAPI Directory `{dir}` not found. "
+                "Please check your `autoapi_dirs` setting.".format(dir=_dir)
+            )
+
+    # Change from app.confdir to app.srcdir.
+    # Before:
+    # - normalized_root = os.path.normpath(
+    # -    os.path.join(app.confdir, app.config.autoapi_root)
+    # -)
+    normalized_root = os.path.normpath(os.path.join(app.srcdir, app.config.autoapi_root))
```
Comment (mik-laj) [inline docs/exts/docs_build/run-patched-sphinx.py]: Here is the root of the problems.  `app.confdir` > `app.srcdir`

**Comment by potiuk** [comment]
Sounds promising :)

File: docs/exts/docs_build/docs_builder.py [hunk 2]
```diff
@@ -255,69 +190,44 @@ def check_spelling(self, verbose: bool, dockerized: bool) -> List[SpellingError]
...
                     warning_text += spelling_file.read()
 
-            spelling_errors.extend(parse_spelling_warnings(warning_text, self._src_dir, dockerized))
+            spelling_errors.extend(parse_spelling_warnings(warning_text, self._src_dir))
             console.print(f"[blue]{self.package_name:60}:[/] [red]Finished spell-checking with errors[/]")
         else:
             if spelling_errors:
                 console.print(
-                    f"[blue]{self.package_name:60}:[/] [yellow]Finished spell-checking " f"with warnings[/]"
+                    f"[blue]{self.package_name:60}:[/] [yellow]Finished spell-checkingwith warnings[/]"
```
Comment (potiuk) [inline docs/exts/docs_build/docs_builder.py]:
```suggestion
                    f"[blue]{self.package_name:60}:[/] [yellow]Finished spell-checking with warnings[/]"
```

**Comment by potiuk** [comment]
Assuming it all works, I think we should make one more change - I think the --jobs args should default to 0. this will significantly speed up docs building for an average user who have more than 2 CPUS without any special action from the user side.

we could also remove the "sequentially" part altogether. Running wth -j 1 would be equivalent of it (with different logging, but I think for an average user it is still perfectly ok to see the output until error occurs.

**Review by kaxil — APPROVED** [review]

**Review by TobKed — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Closed and re-opened to trigger the build again. @mik-laj - If you want, I can remove the parallel/sequential dualism and set -j 0 as default in a follow-up PR.

**Comment by potiuk** [comment]
Hey @mik-laj - FYI.  I rebased / force-pushed this one, as it was based on a commit which was failing (lack of `blinker`)

**Comment by mik-laj** [comment]
I fixed the static checks errors and rebased to latest master.  I hope I can merge this soon and then we can make other improvements. I would like to improve the stability a bit, because by introducing parallel building now building is less stable, e.g. docs packakes without inventories should be built first (see: https://github.com/apache/airflow/blob/e49722859b81cfcdd7e4bb8e8aba4efb049a8590/docs/build_docs.py#L516)

**Review by potiuk — CHANGES_REQUESTED** [review]
@mik-laj  - this one takes a looong time to build > 25 minutes. The docker version is < 10 minutes. Are you sure it's working fine ?

**Comment by potiuk** [comment]
It looks like the build is sequential

File: scripts/ci/docs/ci_docs.sh [hunk 3]
```diff
@@ -22,16 +22,4 @@ build_images::prepare_ci_build
...
-export PYTHONPATH=${AIRFLOW_SOURCES}
-
-pip install --upgrade pip==20.2.4
-
-pip install .[doc] --upgrade --constraint "${AIRFLOW_SOURCES}/constraints.txt"
-
-start_end::group_end
-
-"${AIRFLOW_SOURCES}/docs/build_docs.py" -j 0 "${@}"
+runs::run_docs "${@}"
```
Comment (potiuk) [inline scripts/ci/docs/ci_docs.sh:25]: Yes @mik-laj - I think -j 0 is missing. I think you should set the jobs parameter to be 0 by default.

Comment (mik-laj) [inline scripts/ci/docs/ci_docs.sh:25]:
```suggestion
runs::run_docs -j 0 "${@}"
```

**Comment by potiuk** [comment]
I think it's better to change the --jobs default value to be 0 actually

**Comment by potiuk** [comment]
❤️

**Comment by mik-laj** [comment]
There seems to be some problem. I am looking at it.

**Comment by potiuk** [comment]
Yeah :(

**Comment by mik-laj** [comment]
I pushed the fix. In addition, I noticed that multiprocessing does not work on Mac OS and I fixed it.

**Comment by mik-laj** [comment]
"Build docs" step is greeen.

Before:

> Elapsed time spent in the script: 589 seconds

After:

> Elapsed time spent in the script: 510 seconds

~15% faster.

**Comment by potiuk** [comment]
I pushed rebase for master that likely casued the kubernetes tests failures

**Review by potiuk — APPROVED** [review]

**Comment by ashb** [comment]
@mik-laj  Can you please report the AutoAPI src dir issue upstream?

**Comment by potiuk** [comment]
I believe @mik-laj works on moving to newer version of autoapi #15206 and I think it's either solved there already or will be done as part of this work.

**Comment by ashb** [comment]
Ah cool.

**Pull request merged**